File: java/src/main/java/com/workshop/architecture/fitness/clean_architecture/entity/MembershipBillingReference.java
package com.workshop.architecture.fitness.clean_architecture.entity;

import org.jspecify.annotations.NonNull;

import java.time.Instant;
import java.time.LocalDate;
import java.util.UUID;

public record MembershipBillingReference(UUID id,
                                         UUID membershipId,
                                         String externalInvoiceId,
                                         String externalInvoiceReference,
                                         LocalDate dueDate,
                                         String status,
                                         Instant createdAt,
                                         Instant updatedAt) {

    public static @NonNull MembershipBillingReference create(Membership membership, String externalInvoiceId, MembershipInvoiceDetails membershipInvoiceDetails) {
        var now = Instant.now();
        return new MembershipBillingReference(
                UUID.randomUUID(),
                membership.id(),
                externalInvoiceId,
                membershipInvoiceDetails.membershipId().toString(),
                membershipInvoiceDetails.dueDate(),
                "OPEN",
                now,
                now
        );
    }
}

File: python/src/workshop_api/fitness/clean_architecture/entity/membership_billing_reference.py
import uuid
from dataclasses import dataclass
from datetime import UTC, date, datetime
from uuid import UUID

from workshop_api.fitness.clean_architecture.entity.membership import Membership
from workshop_api.fitness.clean_architecture.entity.membership_invoice_details import (
    MembershipInvoiceDetails,
)


@dataclass(frozen=True)
class MembershipBillingReference:
    id: UUID
    membership_id: UUID
    external_invoice_id: str
    external_invoice_reference: str
    due_date: date
    status: str
    created_at: datetime
    updated_at: datetime

    @staticmethod
    def create(
        membership: Membership,
        external_invoice_id: str,
        membership_invoice_details: MembershipInvoiceDetails,
    ) -> MembershipBillingReference:
        now = datetime.now(UTC)
        return MembershipBillingReference(
            id=uuid.uuid4(),
            membership_id=membership.id,
            external_invoice_id=external_invoice_id,
            external_invoice_reference=str(membership_invoice_details.membership_id),
            due_date=membership_invoice_details.due_date,
            status="OPEN",
            created_at=now,
            updated_at=now,
        )
